src/lib.rs

```diff
@@ -521,20 +520,27 @@ fn to_hex<'a>(src: &[u8], target: &'a mut [u8]) -> Result<&'a str, ()> {
     return unsafe { Ok(str::from_utf8_unchecked(result)) };
 }
 
+#[cfg(feature = "rand")]
+pub(crate) fn random_32_bytes<R: rand::Rng + ?Sized>(rng: &mut R) -> [u8; 32] {
+    let mut ret = [0u8; 32];
+    rng.fill(&mut ret);
+    ret
+}
+
 #[cfg(test)]
 mod tests {
-    use std::marker::PhantomData;
+    #[allow(unused_imports)]    // When building with no default features.
+    use super::*;
+
     use std::str::FromStr;
 
-    use rand::{thread_rng, RngCore};
     #[cfg(target_arch = "wasm32")]
     use wasm_bindgen_test::wasm_bindgen_test as test;
 
-    use crate::context::*;
-    use crate::ffi::types::AlignedType;
-    use crate::ffi::{self};
+    #[cfg(feature = "alloc")]
+    use crate::{ffi, PublicKey, Secp256k1, SecretKey};
```

Comment:
IIUC these do not require `alloc` but only tests requiring `alloc` happen to use them, right?

It's not very readable but the only somewhat reasonable idea beyond adding a comment that comes to my mind is having `test_std`, `test_alloc`, `test_rand`... etc modules and the modules have appropriate cfg. Maybe not having same cfgs multiple times is good enough benefit for it?

Reply:
I think this is a pretty common pattern in test modules to have feature-gated imports because there are feature-gated tests.